### PR TITLE
CP-47001: stdext: quickcheck-style tests and select->epoll conversion

### PR DESCRIPTION
**Pull request by edwintorok**
This is a PR on top of https://github.com/xapi-project/xen-api/pull/5378, we need to merge that first and then the history here will be a lot smaller.

Previous PR which had some comments:
https://github.com/xapi-project/stdext/pull/80
https://github.com/xapi-project/stdext/pull/83

I'll try to address those comments once I've done some more XenRT testing on this branch

### CODE REVIEW

**Comment by edwintorok** [comment]
I fixed some bugs in the `proxy` code, and added some tests for it. Running some tests on this branch now.

**Comment by edwintorok** [comment]
A XenRT BST of this was green, but given all the churn on master now I'll retest (and also do a bit more wider testing), I'm marking this as ready to review meanwhile though.

**Review by psafont — APPROVED** [review]
Why is the library placed in `lib/` isntead of `ocaml/libs`? Does this PR mean that the Resources package and its unixfd are deprecated?

**Comment by edwintorok** [comment]
> Why is the library placed in `lib/` isntead of `ocaml/libs`? 

It got moved from a PR in the `stdext` repo, where the library was in `lib`. Git was smart enough to move most things across to `ocaml/libs` when cherry-picking, but not brand new files. Looks like I'll have to move those.

> Does this PR mean that the Resources package and its unixfd are deprecated?

They solve different problems. `Resources`/safefd attempts to track open file descriptors and tries to prevent leaking them at compile time (by providing appropriate wrappers to close/move ownership), whereas this module is a runtime solution to prevent EBADF.
We could probably considerably simplify the resources/safefd library now that everything is in one repo, and we have this xapi-fd module (i.e. provide a thin wrapper on top of it).

I'll put this back into draft for now, so that I make the 'git mv' when I'm back from PTO.

**Comment by edwintorok** [comment]
For now I've done a 'git mv' as the last step (I'll need to see whether I can use git filter-repo to move it to the correct place, a git rebase with git mv just results in conflicts).

**Comment by edwintorok** [comment]
I fixed the conflicts with `feature/perf`.

**Review by mg12 — APPROVED** [review]

File: lib/xapi-fd-test/observations.mli [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+  and calls [f] with it.
+  It observes [f]'s actions from the other side of a pipe, socket, file descriptor,
+  or block device if possible.
+  It expects [size] bytes written by [f].
+
+  @returns an observation of [f]'s actions on the file descriptor and [f]'s result
+ *)
+
+val observe_rw :
+     (([> readable], kind) Properties.t, bytes) operation
```
[the remark this answers is not in the document]
Comment (edwintorok) [reply lib/xapi-fd-test/observations.mli]: CI passes (which runs the tests in this branch) and an internal BVT passes too.
Merging this to `feature/perf` now, we still have an opportunity to run (or write were needed) more tests before this reaches `master`.

**Pull request merged**